File: 2. OA-2023/APT-Data/P2.cpp
#include <bits/stdc++.h>
using namespace std;

int n, k, degree[301], deleted[301], seen[301], distnce[301], diameter;
vector<int> G[301];

void getDist(int node, int dist)
{
	seen[node] = 1;
	for(auto i: G[node])	{
		if(not seen[i])	{
			if(not deleted[i])	{
				getDist(i, dist+1);
			}
		}
	}
	distnce[node] = dist;
}

int getDiameter()
{
	memset(distnce, -1, sizeof distnce);
	memset(seen, 0, sizeof seen);
	for(int i=0; i<n; i++)	{
		if(not deleted[i])	{
			getDist(i, 0);
			break;
		}
	}
	int mxd = -1, mxi;
	for(int i=0; i<n; i++)	{
		if(distnce[i] > mxd)	{
			mxd = distnce[i];
			mxi = i;
		}
	}
	// cout << mxi << endl;
	memset(distnce, -1, sizeof distnce);
	memset(seen, 0, sizeof seen);
	getDist(mxi, 0);
	return *(max_element(distnce, distnce+n));
}

int calculate(int node)
{
	// cout << "calc " << node << endl;
	memset(distnce, 0, sizeof distnce);
	memset(seen, 0, sizeof seen);
	getDist(node, 0);
	int ret = 0;
	for(int i=0; i<n; i++)	
		if(distnce[i] == diameter)
			ret++;
	return ret;
}



int main()
{
	ios::sync_with_stdio(false);
	cin >> n >> k;
	for(int i=0, u, v; i<n-1; i++)	{
		cin >> u >> v;
		--u; --v;
		G[u].push_back(v);
		G[v].push_back(u);
		degree[u]++; degree[v]++;
	}
	memset(deleted, 0, sizeof deleted);
	pair<int, int> best(0, -1);
	memset(seen, 0, sizeof seen);
	diameter = getDiameter();
	// cout << "fuck " << endl;
	for(int removed = 1; removed <= k; removed++)	{
		for(int i=0; i<n; i++)	{
			if(not deleted[i])	{
				if(degree[i] == 1)	{
					int temp = calculate(i);
					if(temp > best.first)
						best = {temp, i};
				}
			}
		}
		deleted[best.second] = 1;
		for(auto i: G[best.second])
			if(not deleted[i])
				degree[i]--;
		best = {0, -1};
		memset(seen, 0, sizeof seen);
		diameter = getDiameter();
	}
	cout << diameter << endl;
}

File: 2. OA-2023/APT-Data/P2.py
n, k = map(int, input().split())
degree = [0] * 1001
deleted = [0] * 1001
seen = [0] * 1001
distance = [0] * 1001
G = [[] for _ in range(1001)]

def get_distance(node, dist):
    seen[node] = 1
    for i in G[node]:
        if not seen[i]:
            if not deleted[i]:
                get_distance(i, dist + 1)
    distance[node] = dist

def get_diameter():
    global distance, seen
    distance = [-1] * 1001
    seen = [0] * 1001
    for i in range(n):
        if not deleted[i]:
            get_distance(i, 0)
            break
    mxd = -1
    mxi = 0
    for i in range(n):
        if distance[i] > mxd:
            mxd = distance[i]
            mxi = i
    distance = [-1] * 1001
    seen = [0] * 1001
    get_distance(mxi, 0)
    return max(distance)

def calculate(node):
    global distance, seen
    distance = [0] * 1001
    seen = [0] * 1001
    get_distance(node, 0)
    ret = 0
    for i in range(n):
        if distance[i] == diameter:
            ret += 1
    return ret

for i in range(n - 1):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    G[u].append(v)
    G[v].append(u)
    degree[u] += 1
    degree[v] += 1

best = (0, -1)
diameter = get_diameter()

for removed in range(1, k + 1):
    for i in range(n):
        if not deleted[i]:
            if degree[i] == 1:
                temp = calculate(i)
                if temp > best[0]:
                    best = (temp, i)

    deleted[best[1]] = 1
    for i in G[best[1]]:
        if not deleted[i]:
            degree[i] -= 1

    best = (0, -1)
    diameter = get_diameter()

print(diameter)